templates/oauth2/oauth2-login.html.twig

```diff
@@ -0,0 +1,144 @@
+                <div class="form-group">
+                    <fieldset>
+                        <legend>{{ 'Insert U2F Key'|xlt  }}</legend>
+                        <div>
+                            <ul>
+                                <li>{{ 'Insert your key into a USB port and click the Authenticate button below.'|xlt  }}</li>
+                                <li>{{ 'Then press the flashing button on your key within 1 minute.'|xlt }}</li>
+                            </ul>
+                        </div>
+                    </fieldset>
+                    <button type="button" id="authutf" class="btn btn-primary btn-save" onclick="doAuth()">{{ 'Authenticate U2F'|xlt  }}</button>
+                    <input type="hidden" name="form_requests" value="{{ u2fRequests|attr }}" />
+                    <input type="hidden" name="user_role" value="api">
+                </div>
+                {% endif %}
+
+                {% if enforce_signin_email %}
+                <div class="form-group">
+                    <input class="form-control" type="hidden" value="{{ user.email|attr }}">
+                </div>
+                {% endif %}
+                <div class="form-group">
+                    <input class="form-control" type="hidden" name="username" value="{{ user.username|attr }}">
+                </div>
+                <div class="form-group">
+                    <input class="form-control" type="hidden" name="password" value="{{ user.password|attr }}">
+                </div>
+                <input class="form-control" type="hidden" name="mfa_type" value="TOTP">
+            {% endif %}
+            <hr />
+            <div class="row">
+                <div class="col-md-12">
+                    <div class="btn-group">
+                        {% if mfaRequired is empty %}
+                            <button type="submit" name="user_role" class="btn btn-outline-primary" value="api">{{"OpenEMR Login"|xlt }} <i class="fa fa-sign-in-alt"></i></button>
+                            {% if patientRoleSupport is not empty %}
+                                <button type="submit" name="user_role" class="btn btn-outline-info" value="portal-api">{{ "Patient Login"|xlt  }} <i class="fa fa-sign-in-alt"></i></button>
+                            {% endif %}
+                        {% endif %}
+                    </div>
+                    <div class="form-check-inline float-right">
+                        <input class="form-check-input" type="checkbox" name="persist_login" id="persist_login" value="1">
+                        <label for="persist_login" class="form-check-label">{{ "Remember Me"|xlt  }}</label>
+                    </div>
+                </div>
+            </div>
+        </form>
+    </div>
+</div>
+<script>
+    function doAuth() {
+        let xl = window.top.xl || function(id) { return id; };
+        var f = document.getElementById("userLogin");
+        var requests = JSON.parse(f.form_requests.value);
```

Comment:
Is it possible for `f.form_requests.value` to be an invalid JSON input? If it is, this could throw an exception and would be safer to wrap it with a try/catch

Reply:
Noted, I don't have a problem with tidying this up.

Reply:
Fixed.